Objects/frameobject.c

```diff
@@ -620,39 +620,46 @@ _PyFrame_GetState(PyFrameObject *frame)
     Py_UNREACHABLE();
 }
 
-static void
-add_load_fast_null_checks(PyCodeObject *co)
+static bool
+op_exists(PyCodeObject *co, int opcode, int oparg, int start, int stop)
 {
-    int changed = 0;
+    // This only works when opcode is a non-quickened form:
+    assert(_PyOpcode_Deopt[opcode] == opcode);
+    int check_oparg = 0;
     _Py_CODEUNIT *instructions = _PyCode_CODE(co);
-    for (Py_ssize_t i = 0; i < Py_SIZE(co); i++) {
-        int opcode = _Py_OPCODE(instructions[i]);
-        switch (opcode) {
-            case LOAD_FAST:
-            case LOAD_FAST__LOAD_FAST:
-            case LOAD_FAST__LOAD_CONST:
-                changed = 1;
-                _Py_SET_OPCODE(instructions[i], LOAD_FAST_CHECK);
-                break;
-            case LOAD_CONST__LOAD_FAST:
-                changed = 1;
-                _Py_SET_OPCODE(instructions[i], LOAD_CONST);
-                break;
-            case STORE_FAST__LOAD_FAST:
-                changed = 1;
-                _Py_SET_OPCODE(instructions[i], STORE_FAST);
-                break;
+    while (start < stop) {
+        int check_opcode = _PyOpcode_Deopt[_Py_OPCODE(instructions[start])];
+        check_oparg |= _Py_OPARG(instructions[start]);
+        if (check_opcode == opcode && check_oparg == oparg) {
+            return true;
+        }
+        if (check_opcode == EXTENDED_ARG) {
+            check_oparg <<= 8;
         }
-        i += _PyOpcode_Caches[_PyOpcode_Deopt[opcode]];
+        else {
+            check_oparg = 0;
+        }
+        start += 1 + _PyOpcode_Caches[check_opcode];
     }
-    if (changed && co->_co_cached != NULL) {
-        // invalidate cached co_code object
-        Py_CLEAR(co->_co_cached->_co_code);
-        Py_CLEAR(co->_co_cached->_co_cellvars);
-        Py_CLEAR(co->_co_cached->_co_freevars);
-        Py_CLEAR(co->_co_cached->_co_varnames);
-        PyMem_Free(co->_co_cached);
-        co->_co_cached = NULL;
+    return false;
+}
+
+static void
+maybe_bind_fast_local(_PyInterpreterFrame *f, int i)
+{
+    PyCodeObject *co = f->f_code;
+    if (f->localsplus[i] == NULL && op_exists(co, LOAD_FAST, i, 0, Py_SIZE(co)))
+    {
+        // We're being overly-cautious here (it's too expensive to run a full
+        // flow analysis), but at least one instruction expects i to be bound.
+        // Rather than crashing (or changing co_code), just bind None instead:
+        const char *e = "cannot leave local name %R unbound (assigning None)";
+        PyObject *name = PyTuple_GET_ITEM(co->co_localsplusnames, i);
+        if (PyErr_WarnFormat(PyExc_RuntimeWarning, 0, e, name)) {
```

Comment:
Is there any way to tell how disruptive these warnings would be? I don't know all the major use cases of these functions.

Reply:
The only warnings raised by our test suite (which does *lots* of weird tracing stuff) were in the tests you wrote specifically to test these edge-cases. So I'm pretty sure the impact will be minimal in practice.